### ISSUE TITLE
Semigroup type constraint propagating through layers of functions 

### ISSUE THREAD

**Issue by nhenin** [composewell/streamly#155]
I have this dependency (Semigroup (stream monad item)) propagating through layers and showing details of implementation that conceptually should not be seen.
I have : 
1 ) a function A that stream element A1 : requires Semigroup (stream monad A1)
2) a function B that stream element B1 : requires Semigroup (stream monad B1)
3) a function C that stream using A and B : requires Semigroup (stream monad B1) Semigroup (stream monad A1) and so on...

So I finish ended up with my highest streaming functions detailing all the implementation details on theirs signatures. Can I fix this issue ? It causes me issues for modularizing my codebase...

**Comment by harendra-kumar** [comment]
If the stream type is polymorphic you need to tell the compiler whether the given type variable supports Semigroup instance before you can use a Semigroup op. There is no way you can avoid that. However, you can use a monomorphic stream type e.g. SerialT or AheadT etc. and the compiler can figure out the instance itself without you specifying it.

**Comment by nhenin** [comment]
can't we say something like Semigroup (stream monad ) and aside Semigroup A1 and in your library if you have all these Semigroup instances you can combine somehow to get the equivalent of Semigroup (stream monad A1) ?

**Comment by harendra-kumar** [comment]
I guess we can use `QuantifiedConstraints` extension available in GHC 8.6 to write something like:

```
class (forall m a. MonadAsync m => Semigroup (t m a)) => IsStream t where
```

This will have to be under a GHC 8.6 preprocessor conditional. All streams are not monads, so we cannot use `Monad (t m)` currently, but once we fix #157 we can use that too.

I do not know any other way, if you have an idea please try it and send a PR if required.
